File: microservices/example_haskell_consumer/librdkafka/C/hello_ffi.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>


void hello_ffi() {
    printf("Hello, FFI\n"); 
}

char* modifed_string(char* input) {
    size_t size = strlen(input) + strlen("[C]: I was received: ") + 1;
    char* result = (char*)malloc(size);
    if (result == NULL) {
        printf("[C ERROR]: can not allocate memory!\n");
        return NULL;
    }
    snprintf(result, size, "[C]: I was received: %s", input);
    printf("[C]: I was received: %s\n", input);
    return result;
}

typedef struct {
    int event_id;
} context_t;

typedef void (*callback_t)(context_t* ctx);

void run_event_loop(callback_t cb) {
    setbuf(stdout, NULL);
    int i = 0;
    while (1) {
        printf("[C]: (pointer callback) generating event %d\n", i);
        context_t* ctx = (context_t*)malloc(sizeof(context_t));
        ctx->event_id = i;
        cb(ctx);
        free(ctx);
        sleep(1);
        i++;
    }
}

File: microservices/example_haskell_consumer/librdkafka/FFI/HelloFFI.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE RecordWildCards #-}

{-# LANGUAGE ForeignFunctionInterface #-}

module FFI.HelloFFI where

import Foreign.C.String
import Foreign.C
import Foreign.Marshal.Alloc
import Foreign.Ptr
import Foreign
import Control.Exception

foreign import ccall "hello_ffi"
  c_helloFFI :: IO ()

foreign import ccall "modifed_string"
  c_modifedString :: CString -> IO CString

modifedString :: String -> IO (Either String String)
modifedString input =
  withCString input (\cstr ->
    bracket (c_modifedString cstr) free (\ptr ->
      if ptr == nullPtr
        then return (Left "[FFI ERROR]: was null pointer")
        else try (peekCString ptr) >>= (\result ->
          case result of
            Left (e :: SomeException) -> return (Left (show e))
            Right str -> return (Right str)
          )
      )
    )

data Context = Context
  { eventId :: CInt
  }

instance Storable Context where
  sizeOf :: Context -> Int
  sizeOf _ = sizeOf (undefined :: CInt)

  alignment :: Context -> Int
  alignment _ = alignment (undefined :: CInt)

  peek :: Ptr Context -> IO Context
  peek ptr = do
    eid <- peekByteOff ptr 0
    return (Context eid)

  poke :: Ptr Context -> Context -> IO ()
  poke ptr (Context eid) =
    pokeByteOff ptr 0 eid

type Callback = Ptr Context -> IO ()

foreign import ccall "wrapper"
  mkCallback :: Callback -> IO (FunPtr Callback)

foreign import ccall safe "run_event_loop"
  c_runEventLoop :: FunPtr Callback -> IO ()

callback :: Callback
callback ptr =
  handle 
  (\(e :: SomeException) -> putStrLn $ "[HASKELL EXCEPTION]: " ++ show e)
  (peek ptr >>= (\ctx -> putStrLn $ "[HASKELL]: got event: " ++ show (eventId ctx)))
